### ISSUE TITLE
Valgrind warnings in use of p4est

### ISSUE THREAD

**Issue by BarrySmith** [cburstedde/p4est#62]
We are seeing valgrind warnings in the use of p4est -- ftp://ftp.mcs.anl.gov/pub/petsc/nightlylogs/archive/2019/05/26/examples_master_arch-linux-pkgs-valgrind_es.log
It appears that p4est may be using MPI to communicate arrays that has been allocated but not yet initialized. 

Though this is not technically incorrect code it makes using valgrind to find real memory corruption issues more difficult since many of these "false alarm" messages will be missed in with real memory issues.

**Comment by BarrySmith** [comment]
We are seeing valgrind warnings in the use of p4est -- ftp://ftp.mcs.anl.gov/pub/petsc/nightlylogs/archive/2019/05/26/examples_master_arch-linux-pkgs-valgrind_es.log 
It appears that p4est may be using MPI to communicate arrays that has been allocated but not yet initialized. For example: 

`#	> ==13155== Syscall param writev(vector[...]) points to uninitialised byte(s)
#	> ==13155==    at 0xB536F87: writev (writev.c:49)
#	> ==13155==    by 0xAD5E180: MPL_large_writev (mpl_sock.c:32)
#	> ==13155==    by 0xAD3C098: MPIDI_CH3I_Sock_writev (sock.c:2934)
#	> ==13155==    by 0xAD407EE: MPIDI_CH3_iSendv (ch3_isendv.c:95)
#	> ==13155==    by 0xACFC757: MPIDI_CH3_EagerContigIsend (ch3u_eager.c:566)
#	> ==13155==    by 0xAD12E5B: MPID_Isend (mpid_isend.c:138)
#	> ==13155==    by 0xAB24C5F: PMPI_Isend (isend.c:127)
#	> ==13155==    by 0xC8E00BE: p4est_balance_ext (p4est.c:2105)
#	> ==13155==    by 0xC8E0C2B: p4est_balance (p4est.c:1240)
#	> ==13155==    by 0x58CAC1B: DMSetUp_p4est (pforest.c:1084)
#	> ==13155==    by 0x5C2BB00: DMSetUp (dm.c:821)
#	> ==13155==    by 0x58BCA1D: DMPforestGetPlex (pforest.c:4520)
#	> ==13155==    by 0x58BD606: DMCreateCoordinateDM_p4est (pforest.c:4905)
#	> ==13155==    by 0x5C59CB5: DMGetCoordinateDM (dm.c:5659)
#	> ==13155==    by 0x413489: SetupAuxDM (ex12.c:744)
#	> ==13155==    by 0x413489: SetupDiscretization (ex12.c:790)
#	> ==13155==    by 0x413489: main (ex12.c:959)
#	> ==13155==  Address 0xfda0c89 is 25 bytes inside a block of size 64 alloc'd
#	> ==13155==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
#	> ==13155==    by 0xCB79593: sc_malloc_aligned (sc.c:373)
#	> ==13155==    by 0xCB7961B: sc_malloc (sc.c:524)
#	> ==13155==    by 0xCB7971B: sc_realloc (sc.c:589)
#	> ==13155==    by 0xCB7C0C2: sc_array_resize (sc_containers.c:242)
#	> ==13155==    by 0xC8E9195: sc_array_push_count (sc_containers.h:495)
#	> ==13155==    by 0xC8E9195: sc_array_push (sc_containers.h:512)
#	> ==13155==    by 0xC8E9195: p4est_quadrant_array_push (p4est.h:472)
#	> ==13155==    by 0xC8E9195: p4est_tree_compute_overlap (p4est_algorithms.c:1084)
#	> ==13155==    by 0xC8DC557: p4est_balance_response (p4est.c:1216)
#	> ==13155==    by 0xC8E0006: p4est_balance_ext (p4est.c:2078)
#	> ==13155==    by 0xC8E0C2B: p4est_balance (p4est.c:1240)
#	> ==13155==    by 0x58CAC1B: DMSetUp_p4est (pforest.c:1084)
#	> ==13155==    by 0x5C2BB00: DMSetUp (dm.c:821)
#	> ==13155==    by 0x58BCA1D: DMPforestGetPlex (pforest.c:4520)
#	> ==13155==    by 0x58BD606: DMCreateCoordinateDM_p4est (pforest.c:4905)
#	> ==13155==    by 0x5C59CB5: DMGetCoordinateDM (dm.c:5659)
#	> ==13155==    by 0x413489: SetupAuxDM (ex12.c:744)
#	> ==13155==    by 0x413489: SetupDiscretization (ex12.c:790)
#	> ==13155==    by 0x413489: main (ex12.c:959)`

Though this is not technically incorrect code it makes using valgrind to find real memory corruption issues more difficult since many of these "false alarm" messages will be missed in with real memory issues.

**Comment by cburstedde** [comment]
Let me have a look. An --enable-valgrind-clean configure flag might be in order.

**Comment by cburstedde** [comment]
I tried to wget your log file, but didn't find it. The link might have expired? Is there a recent one?

**Comment by BarrySmith** [comment]
ftp://ftp.mcs.anl.gov/pub/petsc/nightlylogs/archive/2019/08/19/examples_master_arch-linux-pkgs-valgrind_es.log

**Comment by cburstedde** [comment]
> ftp://ftp.mcs.anl.gov/pub/petsc/nightlylogs/archive/2019/08/19/examples_master_arch-linux-pkgs-valgrind_es.log

Thanks a lot.  Can you try the feature-valgrind-clean branch?  I've initialized a bunch of unused pad bytes.

**Comment by tisaac** [comment]
Reviving this:

Thanks Carsten, I'm merging this into the version of p4est that petsc uses.  Once we've verified this,
could we come out with a point release soon so that petsc users can rely on a tarball instead of autotools?

**Comment by cburstedde** [comment]
> Reviving this:
> 
> Thanks Carsten, I'm merging this into the version of p4est that petsc uses.  Once we've verified this,
> could we come out with a point release soon so that petsc users can rely on a tarball instead of autotools?

Yes this will be fine.  This branch also contains an update for MPI I/O mesh
loading.  Please let us know how it goes and if there's anything else we might
address short-term.

**Comment by cburstedde** [comment]
Hello, any news? My goal is to have prev3-develop work with all external libraries.

**Comment by cburstedde** [comment]
I'll try to release v2.8 fairly soon, which will be the version without fortran, blas, lapack and with those fixes. Sorry it is taking a while!

**Comment by cburstedde** [comment]
Hi @tisaac, v2.3.2 has these updates integrated.

**Comment by cburstedde** [comment]
Ok so how does the latest v2.8 look for y'all?
